github webhook: Support all actions on discussion events.

Niloth-p opened the pull request:
Fixes: #26064.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Screenshots and screen captures:</summary>

![answered](https://github.com/user-attachments/assets/58c33471-d4e5-429b-acaa-9dcb56319ed0)
![category_changed](https://github.com/user-attachments/assets/07327aa9-6013-47ac-9cb6-cddda60b37a2)
![closed](https://github.com/user-attachments/assets/0c2bf9f3-5cc9-41bd-8cfa-c34e579d9ae0)
![created](https://github.com/user-attachments/assets/94188e4d-a1b0-4f29-bb73-a7994682ca7a)
![deleted](https://github.com/user-attachments/assets/97d42797-0d27-4ee6-9d71-c3454368d8f4)
![edited_body](https://github.com/user-attachments/assets/f5367338-3237-472b-a782-694174471c91)
![edited_title](https://github.com/user-attachments/assets/67cb2bc2-8db6-412f-9d4f-1aeda01dd25f)
![labeled](https://github.com/user-attachments/assets/10e10192-bc50-4c6d-b58a-17b79826c1ae)
![locked](https://github.com/user-attachments/assets/fb89ee4a-4bbb-453b-80b7-0c83097d53a2)
![pinned](https://github.com/user-attachments/assets/b55b566c-02b4-4333-9bac-a1e55a9e6983)
![reopened](https://github.com/user-attachments/assets/1e9d3808-a580-4496-87d0-11a24acd67ea)
![transferred](https://github.com/user-attachments/assets/8336c7f0-51ed-4ed3-b811-3b6db91f8a15)
![unanswered](https://github.com/user-attachments/assets/966b1ae8-ddff-463c-bdde-ed3cca138c2a)
![unlabeled](https://github.com/user-attachments/assets/4c5f7f38-4a8b-4810-9d86-c26dad20df6a)
![unlocked](https://github.com/user-attachments/assets/16a50630-b8fa-4148-b297-a564198b81df)
![unpinned](https://github.com/user-attachments/assets/68a7f3d3-511d-4135-ab79-3f45e2f5ad51)

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p commented on zerver/webhooks/github/view.py:
```diff
@@ -302,16 +317,77 @@ def get_push_commits_body(helper: Helper) -> str:
+        raise KeyError(f"Key '{key}' not found in LazyContext.")
+
+
 def get_discussion_body(helper: Helper) -> str:
```
The other approach would probably involve writing a bunch of `get_discussion_*_body()` functions with the templates inside them, and adding them to the `EVENT_FUNCTION_MAPPER`.
But, I thought this way, we have 2 levels of mapping - 1 at the event level, and 1 at the action level.
And we get to reduce the redundancy of passing the same fields to format() across functions.
We already have a ton of tiny functions in github/view.py, this way we end up with just 1 extra function, a class and a dict constant, and do not add to the pollution.

alya commented:
How are the images you generated being used? At least some screenshots to illustrate what's going on would be very helpful.

Niloth-p commented:
@alya The images are not being used anywhere.
I only added them for making it easy to review the message templates that are being added. They're not meant to be merged.

This PR adds messages for discussion events from GitHub.
So, the screenshots show how the message gets formatted as, for each type.

alya commented:
OK, if they aren't meant to be merged, then please remove them from the PR. You can put them in the PR description.

alya commented:
Oh, I see, this was not clear to me from the PR description.

alya commented:
In the "deleted" notification, does the link actually work?

alya commented on static/images/integrations/github/transferred.png:
I think "webhook-tester" should also be bold?

alya replied:
Is the "discussion #3" link useful here, or should we just make it bold?

alya replied:
This one also looks to be missing a period at the end.

Niloth-p replied:
I've removed both of the expired links.

I've made the previous repo name bold, but as long as the user does not provide a specific topic for all their events, the repo name will be part of the topic name, so I hadn't previously emphasized it. And we currently don't change the topic name on a discussion being moved.

alya commented:
The screenshots look reasonable to me other than those quick notes.

Niloth-p reviewed (commented):
Nice catches, thank you for the review, @alya.

Yes, I was planning to give you better info when formally requesting the review in my check-in, than just mentioning the PR number out of the blue, my bad.

Interestingly, most of the message templates for the GitHub integration do not use bold formatting. Just about 5 of them emphasize user names, among the tens of templates in `git.py`.
Would we want to update them?

Niloth-p commented:
@zulipbot add "integration review"

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -302,16 +317,77 @@ def get_push_commits_body(helper: Helper) -> str:
+            "author": lambda: self.payload["discussion"]["user"]["login"].tame(check_string),
+            "url": lambda: self.payload["discussion"]["html_url"].tame(check_string),
+            "action": lambda: self.payload["action"].tame(check_string),
+            "optional_title": lambda: f' {self.template_values["title"]()}'
```
`configured_title` seems clearer?

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -302,16 +317,77 @@ def get_push_commits_body(helper: Helper) -> str:
+            "old_category": lambda: self.payload["changes"]["category"]["from"]["name"].tame(
+                check_string
+            ),
+            "locked_reason": lambda: f' as {self.payload["discussion"]["active_lock_reason"].tame(check_string)}'
```
Makes sense. Put this in a comment above this line -- it's useful information for all future readers of the line of code too, so belongs in a code comment, not a GitHub one.

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -302,16 +317,77 @@ def get_push_commits_body(helper: Helper) -> str:
+            if self.payload["discussion"]["active_lock_reason"]
+            else "",
+            "closed_reason": lambda: self.payload["discussion"]["state_reason"].tame(check_string),
+            "answer_field": lambda: "old_answer"
```
Again, this probably belongs in a code comment, maybe with more nouns, like "answer_field is purely ...".

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -302,16 +317,77 @@ def get_push_commits_body(helper: Helper) -> str:
     )
 
 
+class LazyContext(dict[str, str | int]):
```
I think the longer-term direction @andersk had in mind for this sort of thing would be to be declaring Pydantic types that spell out each event type with the fields we intend to parse. I'm not sure how practical that is for this huge number of event types involved here with different types without some work on the framework, though; this seems OK for now.

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -302,16 +318,75 @@ def get_push_commits_body(helper: Helper) -> str:
+    action = get_discussion_action(payload)
+    DISCUSSION_TEMPLATE = DISCUSSION_TEMPLATES[action]
+    context = LazyContext(payload, helper.include_title)
+    return DISCUSSION_TEMPLATE.format_map(context)
```
How do discussion events show up in the set of "event type" options for filtering in the settings UI? I'm not sure if it'll just be a single "discussion" item or one per each of these dozen subtypes.

Niloth-p replied:
> How do discussion events show up in the set of "event type" options for filtering in the settings UI?

There's just 1 "event" `discussion` and all these "actions" within that event. So, just 1 entry.
While 16 entries just for the discussion event would pollute the list, not being able to opt out of particular actions is inconvenient indeed.
Would it be worth extending event filtering to include one or more sub-event layers?

timabbott replied:
Yeah, I'm not sure what's best, maybe we should just leave it as one cluster and see what feedback we get (if any) on this detail.

Niloth-p commented:
Updates:
Added 2 comments - for `locked_reason` and `answer_field`.
Renamed `optional_title` to `configured_title`.
Removed the `TODO` comment, will open an issue later to track it.

timabbott commented on zerver/webhooks/github/tests.py:
```diff
@@ -611,13 +613,80 @@ def test_team_edited_with_unsupported_keys(self) -> None:
+        self.check_webhook("discussion__unlocked", TOPIC_DISCUSSION, expected_message)
+
+    def test_discussion_unpinned(self) -> None:
+        expected_message = "Niloth-p **unpinned** [discussion #3](https://github.com/Niloth-p/webhook-tester/discussions/3)."
```
Is the use of bold for these something that we have as an existing standard in this integration?

I don't see any existing GitHub webhook events that do that -- they may bold the name, but never the action. 

I think it'd be better to be consistent with the existing styling? I think the bold would be quite distracting in practice.

timabbott replied:
Great.

timabbott commented:
Merged, thanks for all the work on this @Niloth-p!
